### PR TITLE
Direct access to handlers/state

### PR DESCRIPTION
**Pull request by serkm**
Alternative to #636.
The meta-stack now stays movable and only prompts have a stable address.

### CODE REVIEW

File: libraries/llvm/rts.ll [hunk 1]
```diff
@@ -90,7 +93,7 @@
 %Neg = type {ptr, %Object}
 
 ; Reference into an arena (prompt -- cast to 32bit, offset 32bit)
-%Reference = type {i32, i32}
+%Reference = type { %Prompt, i64 }
```
Comment (b-studios) [inline libraries/llvm/rts.ll:96]: @serkm the comment looks outdated

File: libraries/llvm/rts.ll [hunk 2]
```diff
@@ -442,58 +410,94 @@ define private %Stack @reset(%Stack %oldStack) {
...
-    ret %Stack %uniqueStack
+    ret %Stack %stack
 }
 
-define private void @resumeUnique(%Stack %stack, %Stack %oldStack) {
-    %stackRest = getelementptr %StackValue, %Stack %stack, i64 0, i32 4
-    %rest = load %Stack, ptr %stackRest
-    %isNull = icmp eq %Stack %rest, null
-    br i1 %isNull, label %done, label %next
+define private void @updatePrompts(%Stack %stack) alwaysinline {
```
Comment (b-studios) [inline libraries/llvm/rts.ll:428]: Is this `alwaysinline` here important?

Comment (b-studios) [reply libraries/llvm/rts.ll:428]: I think it doesn't matter since it is only used in `resume`, right?

Comment (serkm) [reply libraries/llvm/rts.ll]: Yes, it doesn't really matter.

File: effekt/shared/src/main/scala/effekt/machine/Transformer.scala [hunk 3]
```diff
@@ -237,9 +237,12 @@ object Transformer {
               // TODO currently we use prompt 1 as a quick fix...
               //    However, this will not work when reinstalling a fresh stack
               //    We need to truly special case global memory!
```
Comment (phischu) [inline effekt/shared/src/main/scala/effekt/machine/Transformer.scala:239]:
```suggestion
```

File: effekt/shared/src/main/scala/effekt/generator/llvm/Transformer.scala [hunk 4]
```diff
@@ -508,6 +511,7 @@ object Transformer {
 
   def transform(value: machine.Variable)(using FunctionContext): Operand =
     substitute(value) match {
+      case machine.Variable("global", machine.Type.Prompt()) => ConstantGlobal("global")
```
Comment (phischu) [inline effekt/shared/src/main/scala/effekt/generator/llvm/Transformer.scala:514]:
```suggestion
      // TODO rethink existence of global
      case machine.Variable("global", machine.Type.Prompt()) => ConstantGlobal("global")
```

File: effekt/shared/src/main/scala/effekt/machine/Transformer.scala [hunk 5]
```diff
@@ -237,9 +237,12 @@ object Transformer {
...
-              LiteralInt(prompt, 1L,
-                Allocate(reference, value, prompt,
-                  transform(body)))
+              // LiteralInt(prompt, 1L,
+              //   Allocate(reference, value, prompt,
+              //     transform(body)))
```
Comment (phischu) [inline effekt/shared/src/main/scala/effekt/machine/Transformer.scala:242]:
```suggestion
```

File: libraries/llvm/rts.ll [hunk 6]
```diff
@@ -741,10 +801,15 @@ define private void @topLevelEraser(%Environment %environment) {
     ret void
 }
 
+@global = private global { i64, %Stack } { i64 0, %Stack null }
+
 define private %Stack @withEmptyStack() {
     ; TODO all stacks share the same source of fresh prompts
```
Comment (phischu) [inline libraries/llvm/rts.ll:807]:
```suggestion
```

**Review by phischu — APPROVED** [review]
Beautiful!

**Pull request merged**